### PR TITLE
Improvement/cldsrv 436 bp conds

### PR DESCRIPTION
**Pull request by KazToozs**
### Motivation and context

Adds Bucket Policy condition handling for limiting IPs and Object Lock retention time.

### Related issues

This PR is based on and dependant of code in CLDSRV-431.

### CODE REVIEW

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (jonathan-gramain) [inline lib/api/apiUtils/authorization/permissionChecks.js]:
```suggestion
            for (const conditionOperator of conditionOperators) {
```

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (jonathan-gramain) [inline lib/api/apiUtils/authorization/permissionChecks.js]: The comment could be more instructive explaining what is the AWS behavior that we reproduce here, if it is not obvious, e.g.

```suggestion
                // AWS returns an error if the condition key doesn't start with `aws:`, so we reproduce this behavior here
```

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (jonathan-gramain) [inline lib/api/apiUtils/authorization/permissionChecks.js]:
```suggestion
                if (validCondition && validCondition.conditionValueTypeChecker) {
                    const conditionValueTypeError = validCondition.conditionValueTypeChecker(conditionValue);
                    if (conditionValueTypeError) {
                        return conditionValueTypeError;
                    }
                }
```

File: lib/api/bucketPutPolicy.js
Comment (jonathan-gramain) [inline lib/api/bucketPutPolicy.js]: Do we need an extra step in the waterflow? I think we could just add the `validatePolicyConditions` block right after the `validatePolicyResource` one above.

Comment (KazToozs) [reply lib/api/bucketPutPolicy.js]: More like [this](https://github.com/scality/cloudserver/pull/5501/files#diff-f9ed4547634b44b1019792bc2c7bdd3fd5c512d9e0f6ed3ba3c67b62b8745efaR69)?

**Review by jonathan-gramain — APPROVED** [review]

**Review by benzekrimaha — APPROVED** [review]
LGTM

**Review by benzekrimaha — CHANGES_REQUESTED** [review]
blocking this PR until CI passing on 8.x

File: package.json
Comment (ghost) [inline package.json]: Adding another blocker for this PR as this Arsenal version is not a valid tag, let's merge it first

**Review by ghost — CHANGES_REQUESTED** [review]

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js:237]: Not for this PR, but maybe worth testing the IP bindings with a fully deployed product (Integration, Zenko), and ensure we are properly passing and accessing the IP from behind a reverse proxy (nginx).

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js]: Performance-wise this may not be the best way to check if the IP is valid. When I worked on the product performance, IP checks indeed had performance impacts, but this is something we do already in the ` IpCheck.ts`  file from Arsenal. 

You will have access to the `ipaddr.IPv4.isValid` function, as well as the `ipaddr.IPv6.isValid` one, as we should support both ipv6 and ipv4. ALso if would make more sense to return a boolean for such a function (see https://github.com/scality/Arsenal/blob/development/7.70/lib/ipCheck.ts#L36)

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js:521]: We will need some unit tests for this function to make sure we are not crashing under specific cases

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js:523]: From what I can see here, the checker is the above function, that only checks ipv4 addresses without much logic (e.g., the CIDR can be 99 and be accepted, while it's an invalid value). So I am not sure this checker function will be enough, wile relying on the helpers we have in Arsenal will make it robust.

Noite: in Arsenal, in `conditions.ts`, we already support conditions checks for IP addresses, with NotIpAddress and IpAddress cases.-

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js:559]: Same here, as we return a string and end up returning the `InvalidPolicyDocument` error, we can just return booleans in this function.

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [reply lib/api/apiUtils/authorization/permissionChecks.js]: Example from AWS here: https://docs.aws.amazon.com/marketplace/latest/buyerguide/buyer-security-ipv6-upgrade.html

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js:546]: While standard, is it something we want to support, as this will have no effect?
I am thinking about very long condition keys; as the BP is stored in the bucket MD, we want to avoid using all the available space for th BP (yes, this is to protect against some attacks, not the nominal use case :slightly_smiling_face: )

**Review by ghost — CHANGES_REQUESTED** [review]

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js:237]: Yes, I believe I cover this in the [integration PR](https://github.com/scality/Integration/pull/1218) you have seen. I'll resolve this conversation, feel free to reopen.

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js:559]: `MalformedPolicy` error is returned upon an error in the policy, with relevant information to what the malformation is, hence I chose to use the abstractions we saw in the comments above to return the valid error string per error. Do you suggest doing this in another way?

File: package.json
Comment (KazToozs) [reply package.json]: updated 418144caa4a30dae305e46c66802550efd29dc34

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js]: Using the arsenal function here: ad0d3adaa8432feb3d73676cae53fabd088bd423

maintaining the structure to return error message per AWs standard; see [this](https://github.com/scality/cloudserver/pull/5501#discussion_r1444759471) comment and let me know what you think

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js:523]: Same as above: 
Using the arsenal function here: ad0d3adaa8432feb3d73676cae53fabd088bd423

maintaining the structure to return error message per AWs standard; see [this](https://github.com/scality/cloudserver/pull/5501#discussion_r1444759471) comment and let me know what you think

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js:546]: There is a conversation to be had here indeed. I have simply been basing myself on AWS behaviour. I'll address it with the team somehow.

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js:521]: I'll wait for another review to confirm on what the function should be returning before addressing this comment

**Comment by KazToozs** [comment]
ping

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js:546]: Ticket created
https://scality.slack.com/archives/C01M77VU8RJ/p1704819662752699?thread_ts=1704792330.505129&cid=C01M77VU8RJ

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js:521]: Confirmed we are sending back errors directly

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js:523]: Added new verifications for IP using ipaddr from arsenal

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js:559]: Addressed: we are sending back the errors directly

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js]: Updated to use checks from arsenal

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js]: I think the sentence is missing some words :slightly_smiling_face: 

```suggestion
    // These preliminary checks are validating the provided
    // ip address against ipaddr.js, the library we use when
    // evaluating IP condition keys. It ensures compatibility,
    // but additional checks are required to enforce the right
    // notation (e.g., xxx.xxx.xxx.xxx/xx for IPv4). Otherwise,
    // we would accept different ip formats, which is not
    // standard in an AWS use case.
```

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js]:
```suggestion
        if (err) {
            return errString;
        }
```

Also I'm not sure the `if` is needed as we always execute this code when we had an error; unless we want to filter out some specific errors, just returning should work

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js]: Maybe worth a constant in the const file, to avoid recreating it?

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js]: Only personal preference but nested functions can be misleading, you can do that:

```suggestion
    const validateRegex = (ip) => {
```

Or create another function outside of `checkIp`

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (ghost) [inline lib/api/apiUtils/authorization/permissionChecks.js]:
```suggestion
    return errString;
```

File: tests/functional/aws-node-sdk/test/bucket/putBucketPolicy.js
Comment (ghost) [inline tests/functional/aws-node-sdk/test/bucket/putBucketPolicy.js:194]: AWS mentions a maximum of 100 years as the max retention period, can we test that as well?
https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock-managing.html#object-lock-managing-retention-limits

Comment (KazToozs) [reply tests/functional/aws-node-sdk/test/bucket/putBucketPolicy.js:194]: I think that's a test for putBucketRetention and putObjectRetention, and not directly related to conditions so is not in the scope of this PR.
To support this point, I was able to add the following bucket policy to an AWS bucket, 'accepting' objects with more than 50000 days.
```
{
            "Sid": "SetRetentionPeriod",
            "Effect": "Allow",
            "Principal": {
                "AWS": "arn:aws:iam::554628645944:user/WillTest"
            },
            "Action": "s3:PutObjectRetention",
            "Resource": "arn:aws:s3:::test-bucket-retention-condition2/*",
            "Condition": {
                "NumericGreaterThan": {
                    "s3:object-lock-remaining-retention-days": "50000"
                }
            }
        }
```

Of note, on testing the following command with retention until `2223`:
```
aws s3api put-object-retention --bucket test-bucket-retention-condition2 --key image1.png --retention Mode=GOVERNANCE,RetainUntilDate=2223-12-23T23:59:59 --profile playground-test-user --bypass-governance-retention
```

It passes, where the AWS interface plainly gives me the following:

> Retain until date
> December 24, 2223, 00:59:59 (UTC+01:00)

It seems this 100 year limit only applies to the default bucket retention configuration, as I indeed cannot set that to more than 100 years. 
I will create a ticket for that, as well as more tests generally for the PutObjectLockConfiguration API as from I see there are not many currently.

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js]: done 98d51fd670f3b01e0876692a64765bd53c8542b6

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js]: done 98d51fd670f3b01e0876692a64765bd53c8542b6

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js]: done 92f719e34fdb39708803e0f792aa4f896757fd96

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js]: done 98d51fd670f3b01e0876692a64765bd53c8542b6

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js]: done 98d51fd670f3b01e0876692a64765bd53c8542b6

File: lib/api/apiUtils/authorization/permissionChecks.js
Comment (KazToozs) [reply lib/api/apiUtils/authorization/permissionChecks.js]: choosing to maintain the nested function as it's only relevant to the scope of checkIp

File: tests/functional/aws-node-sdk/test/bucket/putBucketPolicy.js
Comment (ghost) [reply tests/functional/aws-node-sdk/test/bucket/putBucketPolicy.js:194]: The doc was making me think we were rejected if trying to put a limit higher than 100 years, but if it works, then all good

**Review by ghost — APPROVED** [review]

**Review by benzekrimaha — APPROVED** [review]

**Comment by KazToozs** [comment]
ping

**Comment by KazToozs** [comment]
ping

**Pull request merged**
